### PR TITLE
Include option for 2018 Jetid working points

### PR DESCRIPTION
**Pull request by knash**
#### PR description:

Update to include 2018 jetid version
https://twiki.cern.ch/twiki/bin/view/CMS/JetID13TeVRun2018

This just updates 
PhysicsTools/SelectorUtils/interface/PFJetIDSelectionFunctor.h
and not  
PhysicsTools/NanoAOD/python/jets_cff.py

#### PR validation:

Tested that the triggers for tight and tightlepveto seem to fire correctly and that the efficiency makes sense

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: PhysicsTools/SelectorUtils/interface/PFJetIDSelectionFunctor.h [hunk 1]
```diff
@@ -443,7 +503,7 @@ class PFJetIDSelectionFunctor : public Selector<pat::Jet>  {
     } else if ( quality_ == TIGHT ) {
       set("CHF", 0.0);
```
Comment (perrotta) [inline PhysicsTools/SelectorUtils/interface/PFJetIDSelectionFunctor.h]: Indentation... or clang-format

**Review by perrotta — COMMENTED** [review]
Also a rebase is needed, see conflicting files

**Comment by perrotta** [comment]
Given the changes observed in DQM in the jenkins tests outputs, and not in the reco output themselves, I have the impression that such PFJetIDSelectionFunctor is picked in the JETMET DQM modules, but not in the reco miniAOD ones.
Could you please shortly remind us how this PFJetIDSelectionFunctor enters in the game, and confirm that what reported in the jenkins tests outputs does make any sense?

**Comment by ahinzmann** [comment]
> Given the changes observed in DQM in the jenkins tests outputs, and not in the reco output themselves, I have the impression that such PFJetIDSelectionFunctor is picked in the JETMET DQM modules, but not in the reco miniAOD ones.
> Could you please shortly remind us how this PFJetIDSelectionFunctor enters in the game, and confirm that what reported in the jenkins tests outputs does make any sense?

JetID is not present in RECO/AOD/MiniAOD but all input variables are stored.
JetID is used for cleaned DQM historgrams, thus we expect DQM output in cleaned historgrams to change.
JetID is also used in NanoAOD, see e.g. here:
https://github.com/cms-sw/cmssw/blob/82e3ed4ea0b8756c412d1f56758b7698d717104e/PhysicsTools/NanoAOD/python/jets_cff.py#L53

**Comment by perrotta** [comment]
> JetID is not present in RECO/AOD/MiniAOD but all input variables are stored.
> JetID is used for cleaned DQM historgrams, thus we expect DQM output in cleaned historgrams to change.
> JetID is also used in NanoAOD, see e.g. here:
> https://github.com/cms-sw/cmssw/blob/82e3ed4ea0b8756c412d1f56758b7698d717104e/PhysicsTools/NanoAOD/python/jets_cff.py#L53

Ok, thank you.
So, jetID is recomputed on the spot when needed (e.g. nanoAOD), and also in DQM.

**Comment by perrotta** [comment]
Now:
- Please rebase to get rid of the conflict
- Please apply code-checks
- Please apply code-format updates
- Please apply all possibly remaing comments which are not yet fixed by then

**Comment by perrotta** [comment]
ping @knash @ahinzmann
By the way: do you plan to backport this PR also in 10_6, and maybe also 10_2 for next nanoAOD campaigns?

**Comment by ahinzmann** [comment]
yes, this should go into 10_2 and 10_6 for
1. future reprocessing of 2018 data
2. analyses applying this on the fly for 2018 analyses.

**Comment by ahinzmann** [comment]
@knash let's wrap this up

**Comment by clelange** [comment]
@knash - while you're updating this, could you please also include the PUPPI Jet ID criteria presented and approved today: https://indico.cern.ch/event/830205/contributions/3477878/attachments/1872771/3082418/Puppi_JetID_JMAR_update.pdf (the ones implemented in this PR by now apply to CHS jets)

**Comment by knash** [comment]
Ok, I will add the 2018 PUPPI option

**Comment by perrotta** [comment]
Please @knash let us know about the expected timeline for updating this PR

**Comment by knash** [comment]
@perrotta Should be pretty quick now.  I have the WP added and now I need to test it for various eta regions

**Comment by perrotta** [comment]
@knash, please take care of 
- Code-checks
- Code-format
- Conflicting files
- Comments received

**Comment by perrotta** [comment]
please test
(Thank you Kevin)

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 2]
```diff
@@ -100,7 +105,9 @@
 for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016:
     modifier.toModify( tightJetIdAK8.filterParams, version = "WINTER16" )
     modifier.toModify( tightJetIdLepVetoAK8.filterParams, version = "WINTER16" )
-
+#Nano without modifier?
+run2_nanoAOD_102Xv1.toModify( tightJetIdAK8.filterParams, version = "SUMMER18" )
+run2_nanoAOD_102Xv1.toModify( tightJetIdAK8.filterParams, version = "SUMMER18" )
```
Comment (perrotta) [inline PhysicsTools/NanoAOD/python/jets_cff.py]: Repetition of the previous line: this should be ```tightJetIdLepVetoAK8```, I guess...

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 3]
```diff
@@ -1,6 +1,9 @@
 import FWCore.ParameterSet.Config as cms
 from Configuration.Eras.Modifier_run2_miniAOD_80XLegacy_cff import run2_miniAOD_80XLegacy
 from Configuration.Eras.Modifier_run2_nanoAOD_94X2016_cff import run2_nanoAOD_94X2016
+from Configuration.Eras.Era_Run2_2018_cff import Run2_2018
```
Comment (perrotta) [inline PhysicsTools/NanoAOD/python/jets_cff.py]: ```Run2_2018``` modifier never used

**Comment by perrotta** [comment]
@ahinzmann , now automatic tests do not show differences even in DQM.
This may look normal to me, but it contradicts what you wrote in https://github.com/cms-sw/cmssw/pull/27184#issuecomment-505001118 : could you please check?

**Comment by ahinzmann** [comment]
> @ahinzmann , now automatic tests do not show differences even in DQM.
> This may look normal to me, but it contradicts what you wrote in [#27184 (comment)](https://github.com/cms-sw/cmssw/pull/27184#issuecomment-505001118) : could you please check?

It would affect DQM if we change the definition picked up by DQM:
https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/DQMOffline/JetMET/python/jetAnalyzer_cfi.py#L110

Since only a new ID is added no change in DQM indeed. For Run3 data taking preparation we should evaluate which ID version is most appropriate.

**Comment by perrotta** [comment]
Ok, thank you @ahnzmann : as such it makes sense!
As soon as the fix https://github.com/cms-sw/cmssw/pull/27184#pullrequestreview-260768709 is applied, this may be ready to go, then.
(Personally, I would also better organize spaces in lines 382-597 of PFJetIDSelectionFunctor.h, in order to keep separated the scopes corresponding to the different ```version```'s and improve readability. But this is something that was already as such in the previous code, and I wouldn't push for it).

**Comment by perrotta** [comment]
@knash , please let us know when do you plan to fix https://github.com/cms-sw/cmssw/pull/27184#pullrequestreview-260768709
Besides that, the PR is basically ready to go...

**Comment by knash** [comment]
@perrotta while making the review changes I noticed that the default operating point that was propagated to NanoAOD for AK8 jets is the CHS version, but should probably be the PUPPI version.  There is not really an official recommendation for AK8 jets anyway, and the only change would be for AK8 jets with eta>2.6 (which I am not sure even exist in analyses), but I think it should be changed for consistency anyway.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
> @perrotta while making the review changes I noticed that the default operating point that was propagated to NanoAOD for AK8 jets is the CHS version, but should probably be the PUPPI version. There is not really an official recommendation for AK8 jets anyway, and the only change would be for AK8 jets with eta>2.6 (which I am not sure even exist in analyses), but I think it should be changed for consistency anyway.

I imagine that the difference observable in the jenkins tests for wf 1325.7 (Run2_2017,run2_nanoAOD_94XMiniAODv2) is exactly due to this modification of the default working point for AK8 jets:
image

Can you confirm?

**Comment by knash** [comment]
Hello @perrotta,
There are changes expected for AK8 jets in 2017 with eta>2.6 so I think this makes sense.
Is there a method for confirming this directly?

I think it would be good to have a NanoAOD expert confirm the operating point change too.

**Comment by perrotta** [comment]
> Hello @perrotta,
> There are changes expected for AK8 jets in 2017 with eta>2.6 so I think this makes sense.
> Is there a method for confirming this directly?
> 
> I think it would be good to have a NanoAOD expert confirm the operating point change too.

xpog (aka @peruzzim) will confirm with his signature whether the operating point change is ok.
reco is fine with the current status of this PR, and ready to sign

**Comment by perrotta** [comment]
+1
- New JetID working points made available to users
- Changes for AK8 jets are propagated to the nanoAOD productions 
- Jenkins tests pass and no differences other that those in the nanoAOD workflows are present

**Comment by perrotta** [comment]
@fgolf @peruzzim please have a look and sign, if you think so

**Comment by clelange** [comment]
Just to add in response to https://github.com/cms-sw/cmssw/pull/27184#issuecomment-511719790 that since we now have both PUPPI and CHS JetID recommendations for AK4 jets, the recommendation is that the corresponding AK8 jet collections should use the respective ID, i.e. the PUPPI AK8 jets should use the PUPPI AK4 Jet ID also in NanoAOD.

**Comment by knash** [comment]
@clelange: Good to know!  Then this is indeed the correct change I think.
But I would recommend updating the Twiki with this recommendation (unless I missed it)

**Comment by clelange** [comment]
Thanks @knash - the [twiki](https://twiki.cern.ch/twiki/bin/view/CMS/JetID13TeVRun2018) has been updated. So far, we had only [announced](https://hypernews.cern.ch/HyperNews/CMS/get/jet-algorithms/543.html) this on hypernews.

**Comment by santocch** [comment]
+1

**Comment by perrotta** [comment]
@peruzzim @fgolf could you please scrutinize this PR?

**Comment by peruzzim** [comment]
As you know, for validation we have routines in the nanoAOD repository that test all supported configuration.
The initial description of this PR said that only selection functions would be changed, but then also changes to nanoAOD configuration were added - why? This makes it impossible to properly review things, should really be avoided by developers.

**Comment by peruzzim** [comment]
agreed with @clange: nanoAOD changes should be removed from this PR, then you merge the code in master, 106X and 102X, then we will add it to nanoAOD

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Only change since last reco signature in https://github.com/cms-sw/cmssw/pull/27184#issuecomment-513135066 is that now the new working points are not propagated to nanoAOD any more.
- Jenkins tests pass and show no differences wrt baseline; not even in nanoAOD, as expected

**Comment by fabiocos** [comment]
@santocch you already signed the previous version of the PR before removing nanAOD, so I assume this is ok for you. Please comment otherwise

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Comment by clelange** [comment]
@peruzzim @fgolf I guess this should be back-ported to at least 10_6_X, can you confirm? Also, any other branch?

**Comment by santocch** [comment]
+1

**Pull request merged**